### PR TITLE
[video_player] Add macOS support

### PR DESCRIPTION
**Pull request by stuartmorgan**
Adds macOS support to `video_player`, sharing almost all of the code with iOS.

Notes about changes at a high level:
- macOS does not have `CADisplayLink` (prior to 14, and even there without all the functionality we need), so this adds macOS compilation branches that use the lower-level `CVDisplayLink` instead. Per the TODO, this code should be extracted later to reduce `ifdef`s in what is already a complicated file.
- Adds KVO unregistration on `dealloc` if it wasn't done in `dispose`, since unit tests were crashing on macOS with that.
- Temporarily ifdef's out `publish:` for macOS, with a TODO to re-enable it after the next stable.

Most of https://github.com/flutter/flutter/issues/41688
Once this lands, the app-facing package will be updated to endorse it for macOS.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
The podspec check failure is because the Flutter Pod we lint against is one that we manually publish occasionally just for this, and it hasn't been updated since those APIs were added. I'll need to dig up the instructions on updating it.

**Comment by stuartmorgan** [comment]
> Includes a `pigeon` update since the version currently being used wasn't generating Obj-C files that compile on macOS.
> * This could easily be moved to a prequel PR if that would be preferred.

I realized it would be better to do [a prequel PR for prefixes](https://github.com/flutter/packages/pull/4994), so this will be removed once that lands.

**Comment by stuartmorgan** [comment]
Tests should all be green now, so this is fully ready.

File: packages/video_player/video_player_avfoundation/darwin/Classes/FVPVideoPlayerPlugin.m [hunk 1]
```diff
@@ -53,18 +80,33 @@ @interface FVPVideoPlayer : NSObject <FlutterTexture, FlutterStreamHandler>
 // An invisible AVPlayerLayer is used to overwrite the protection of pixel buffers in those streams
 // for issue #1, and restore the correct width and height for issue #2.
 @property(readonly, nonatomic) AVPlayerLayer *playerLayer;
```
Comment (tarrinneal) [inline packages/video_player/video_player_avfoundation/darwin/Classes/FVPVideoPlayerPlugin.m:82]: could this go under the endif?

**Review by tarrinneal — APPROVED** [review]
TPMITIDF,BIDKETFTH.

Seems fine.

File: packages/video_player/video_player_avfoundation/darwin/Classes/FVPVideoPlayerPlugin.m (hunk 1 above)
Comment (stuartmorgan) [reply packages/video_player/video_player_avfoundation/darwin/Classes/FVPVideoPlayerPlugin.m:82]: I'll move the ifdef'd properties to the end of the property list, similar to what we do with includes.

**Comment by stuartmorgan** [comment]
> Hm, I thought that script change would fix the check_podspecs failure. It looks like I'll need to find another way to force it to get the right version.

And now it's fine even without the cache clearing 🤷🏻  Maybe just a general CDN propagation issue.

**Review by cbracken — APPROVED** [review]
This looks reasonable. This is a bit on the border of `ifdef`s where it might be worth considering macOS/iOS subclasses for the conditionally compiled logic but it's still a relatively small/innocuous amount overall, so seems fine as is.

**Comment by AlexV525** [comment]
> And now it's fine even without the cache clearing 🤷🏻 Maybe just a general CDN propagation issue.

@stuartmorgan I ran into this in CIs, it was trying to find `ios/video_player_avfoundation.podspec` but got nothing.
https://github.com/fluttercandies/flutter_photo_manager/actions/runs/6374503740/job/17299430662?pr=997

**Pull request merged**